karapace/protobuf/schema.py

```diff
@@ -106,6 +109,49 @@ def option_element_string(option: OptionElement) -> str:
     return f"option {result};\n"
 
 
+@dataclasses.dataclass(frozen=True)
+class UsedType:
+    parent_object_qualified_name: str
+    used_attribute_type: str
+
+
+@dataclasses.dataclass(frozen=True)
+class TypeTree:
+    token: str
+    children: List["TypeTree"]
+
+    @property
+    def is_fully_qualified_type(self) -> bool:
+        return len(self.children) == 0
+
+    def represent(self, level=0) -> str:
+        spacing = " " * 3 * level
+        if self.is_full_path_type:
+            return f"{spacing}{self.token}"
+        child_repr = "\n".join(child.represent(level=level + 1) for child in self.children)
+        return f"{spacing}{self.token} -> \n{child_repr}"
+
+    def __repr__(self) -> str:
+        return self.represent()
+
+
+def add_new_type_r(parent_tree: TypeTree, remaining_tokens: List[str]) -> None:
```

Comment:
nit: What does `_r` mean here? Can it be expanded to some word?

Reply:
it stands for `recursive` I don't know if there is a "pythonic" way of defining a function for the recursive side, I know I can define a function inside the `add_new_type` function but I don't really like nested functions. Maybe I can simply rename it as `_add_new_type` since it's a method internal and someone from external isn't expected to call the recursive one, wdyt?

Reply:
the idea was just to provide an external "happy and nice" function without exposing types that make sense only when looking the recursive implementation, that to me it's an implementation detail and should't be exposed, probably the `_` prepending it's the appropiate solution

Reply:
Yeah prepending `_` would be nice, but I'd also expand the `_r` to `_recurse`, not much gained from saving those few characters.